### PR TITLE
send queue: allow sending reactions to local echoes

### LINKED ISSUE #3663: send queue: dependent actions system
Under less-than-optimal network situations, EX fails to send reactions and "eats" them. It'd be cool if the send queue (+retry mechanisms) was used for reaction sending. Benji suggests that it would probably improve in less duplicate code https://matrix.to/#/!iYnZafYUoXkeVPOSQh:matrix.org/$hGjNwhDvoNeK15O6h3DhmO7wlVrzNsdDBBssJ9o4iYQ?via=matrix.org&via=element.io&via=envs.net

### PR DESCRIPTION
**Pull request by bnjbvr**
Fixes #3663.

### CODE REVIEW

**Comment by bnjbvr** [comment]
So, some jobs have been failing here with an error like this:

```
error[E0275]: overflow evaluating the requirement `matrix_sdk::ruma::ruma_events::OriginalSyncMessageLikeEvent<matrix_sdk::ruma::ruma_events::room::redaction::RoomRedactionEventContent>: std::marker::Send`
    |
    = help: consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`multiverse`)
```

We're making a library, so it's a big no-no to add the requirement that *every user* adds the `recursion_limit` to their crate, just because they're using ours. So I digged a bit, and found another solution.

The issue is that when trying to evaluate if a type is `Send` / `Sync`, rustc will do recursive calls for deeply nested types, until it stops doing so because it thinks it may be blowing up the stack.

A workaround that's been explained to me by a rustc friend: it's possible to "break up" the call tree by adding an `unsafe impl Send` (resp. Sync) for one of the types in the tree. I've done that locally for `SyncTimelineEvent`, which always appears in the error traces.

Now, it happens that `SyncTimelineEvent` is `Send` and `Sync` *right now*, but that might not be true again in the future. So I'd really like the compiler to tell me whenever it becomes not `Send` or `Sync`. As a result, I've introduced a new Cargo feature for the `matrix-sdk-common` crate (that contains the `SyncTimelineEvent`), called `test-sync-send`:

- when this Cargo feature is set, the `unsafe impl`s are not defined, and a test will statically assert that this type implements the `Send` and `Sync` bounds. Basically, back to square one: the compiler does the work of proving this specific type (and not users of it) is `Send` and `Sync`. One of the CI jobs will run with this feature.
- when the Cargo feature isn't set, this test is disabled, and the `unsafe impl`s are active, so the compiler can short-cut proving that this type is `Send` and `Sync`, and we don't hit the recursion limit.

This is a bit hacky, and I hope the next solver will do better; in the meanwhile, is it an acceptable tradeoff?

**Comment by bmarty** [comment]
Hello @bnjbvr , testing the branch on EXA.

in online mode:
- send a message. the message is sent.
- switch off the network (airplane mode for instance)
- send a reaction. Nothing happen on the timeline (no local echo) and I get this log:

```
elementx: org.matrix.rustcomponents.sdk.ClientException$Generic: msg=Failed toggling reaction
at org.matrix.rustcomponents.sdk.FfiConverterTypeClientError.read(matrix_sdk_ffi.kt:25559)
at org.matrix.rustcomponents.sdk.FfiConverterTypeClientError.read(matrix_sdk_ffi.kt:25554)
at org.matrix.rustcomponents.sdk.FfiConverter$DefaultImpls.liftFromRustBuffer(matrix_sdk_ffi.kt:225)
at org.matrix.rustcomponents.sdk.FfiConverterRustBuffer$DefaultImpls.liftFromRustBuffer(matrix_sdk_ffi.kt:237)
at org.matrix.rustcomponents.sdk.FfiConverterTypeClientError.liftFromRustBuffer(matrix_sdk_ffi.kt:25554)
at org.matrix.rustcomponents.sdk.FfiConverterTypeClientError.liftFromRustBuffer(matrix_sdk_ffi.kt:25554)
at org.matrix.rustcomponents.sdk.FfiConverterRustBuffer$DefaultImpls.lift(matrix_sdk_ffi.kt:238)
at org.matrix.rustcomponents.sdk.FfiConverterTypeClientError.lift(matrix_sdk_ffi.kt:25554)
at org.matrix.rustcomponents.sdk.FfiConverterTypeClientError.lift(matrix_sdk_ffi.kt:25554)
at org.matrix.rustcomponents.sdk.ClientException$ErrorHandler.lift(matrix_sdk_ffi.kt:25548)
at org.matrix.rustcomponents.sdk.ClientException$ErrorHandler.lift(matrix_sdk_ffi.kt:25547)
at org.matrix.rustcomponents.sdk.Matrix_sdk_ffiKt.uniffiCheckCallStatus(matrix_sdk_ffi.kt:301)
at org.matrix.rustcomponents.sdk.Matrix_sdk_ffiKt.access$uniffiCheckCallStatus(matrix_sdk_ffi.kt:1)
at org.matrix.rustcomponents.sdk.Matrix_sdk_ffiKt.uniffiRustCallAsync(matrix_sdk_ffi.kt:36163)
at org.matrix.rustcomponents.sdk.Timeline.toggleReaction$suspendImpl(matrix_sdk_ffi.kt:19773)
at org.matrix.rustcomponents.sdk.Timeline.toggleReaction(Unknown Source:0)
```

I cannot test adding reaction on not sent message, I need to do some change on the application to be able to do that, but I will let you know what the result of this test.

**Comment by bnjbvr** [comment]
> Note: I use the eventId for the uniqueId parameter [here](https://github.com/matrix-org/matrix-rust-sdk/pull/3749/files#diff-585addf5b1e9b7be30f6de5cf83d63d84d4f901b98ef4524c703e3ebf14392f1R540), this is maybe not correct, let me know. (EDIT 2: this is probably the root cause of the issue below, give me a minute)

@bmarty Yeah, you need to pass the "internal_id()" or "unique_id()" for the item; this was a way to avoid having to deal with passing an `EventTimelineItem`. I'll tweak the documentation later (after review) to make this clearer.

**Comment by bmarty** [comment]
OK, thanks, passing the `unique_id` is solving the issue, and is working well as well when reacted to a non sent event.

I have found a scenario which is failing though:

- online mode
- send a message
- add a reaction, and wait for everything to be synced.
- switch to offline
- redact the reaction
- react again with the same reaction
- switch to online

Nothing should happen, but at the end the reaction is redacted.

See the video below:

ReactionLocalEcho

**Review by Hywan — APPROVED** [review]
Lot of code. Not sure I'm having all the details but it looks good! Well done.

Now, let's see how many persons will react to their own local event…

File: crates/matrix-sdk/src/send_queue.rs [hunk 1]
```diff
@@ -980,19 +988,29 @@ impl QueueStorage {
     }
 }
 
+/// The content of a local echo.
+#[derive(Clone, Debug)]
+pub enum LocalEchoContent {
+    /// The local echo contains an actual event ready to display.
+    Event {
```
[the remark this answers is not in the document]
Comment (bnjbvr) [reply crates/matrix-sdk/src/send_queue.rs:1098]: Yes, that's the idea!

**Comment by bnjbvr** [comment]
@bmarty opened a followup for the issue you've found, since it's unfortunately not trivial to handle. Thanks though!

**Pull request merged**